Feature/aggregations

frenchbread opened the pull request:
### Changes
- Update packages
- Added eslint & babel
- Completely refactored to support elasticsearch aggregations (all parsing is now done on server-side, client-side only renders & watches for filter updates)
- Added filtering by:
  - date range
  - log type
  - granularity
- Added loading/request metrics to https://github.com/apinf/docs/tree/master/docs/develop/emq

### How it looks

![screen shot 2017-03-16 at 14 55 15](https://cloud.githubusercontent.com/assets/2122679/23997086/99414c56-0a58-11e7-8ce2-cec532abfd66.png)

### How fast it loads

- Very fast
- Instantly

### TODOs

- [x] Show metrics for X&Y axis
- [x] Remove background grid
- [x] Hourly/daily/weekly visualization

Closes #6

frenchbread commented:
Please 👀

bajiat commented:
@Nazarah Can you comment from the UX point of view?

brylie commented:
@frenchbread can you give us some clearly defined metrics here? E.g. how many records is the dashboard loading? How much time (in milliseconds) do the following stages take (basically, what does the Chrome timeline contain)?

- data transfer
- ui paint

brylie commented:
Also, what do the performance metrics look like with 10x records? 100x? A rough estimate is fine, so we can get an idea of how this would scale.

frenchbread commented:
### Server side
The server side request to ES takes around `43-63` milliseconds.

<img width="464" alt="screen shot 2017-03-07 at 13 03 31" src="https://cloud.githubusercontent.com/assets/2122679/23654081/cf7beac4-0337-11e7-8104-c8da9c43ff90.png">

### Client side
On client side, the total page load takes around `2-2.5` seconds, and the actual rendering around `300` milliseconds.

![screen shot 2017-03-07 at 13 16 59](https://cloud.githubusercontent.com/assets/2122679/23654223/5d310890-0338-11e7-8aec-5ba59a9ca50e.png)

brylie commented:
Excellent, and how many records are you retrieving? That way we can figure out the relationship between the number of records and performance time.

frenchbread commented:
As of now the dataset that goes to the charts looks like this:

<img width="469" alt="screen shot 2017-03-07 at 13 18 50" src="https://cloud.githubusercontent.com/assets/2122679/23654299/b6d5a5c2-0338-11e7-95eb-bc4025812430.png">

So it does not really matter how much items are there since they are not grouped "manually" (by dc.js) but are pre-aggregated. The only load I can think of can come from selecting the maximum possible date range, e.g. years, decades? In that case the list of records will grow. Right now, by default, around 30 days (depending on a month) are shown.

I can go ahead & populate ES with more sample data, say for last 2+ years, and compare the results.

brylie commented:
Cool, yes. Keep in mind that the aggregation step takes time too.

![screenshot_20170307_132924](https://cloud.githubusercontent.com/assets/17307/23654608/1daf7a38-033a-11e7-8933-b8e8c6116fb0.png)

frenchbread commented:
@brylie 

> Keep in mind that the aggregation step takes time too.

This is what "server-side" metric measures.


Here is a range for all the data in the ES (1st Jan - 3rd Mar): 3069 items total. And it also takes from `40-60` ms to request data.

```
[{"key":"MQT connections over time","values":[{"x":1483308000000,"y":33},{"x":1483394400000,"y":107},{"x":1483480800000,"y":116},{"x":1483567200000,"y":46},{"x":1483653600000,"y":35},{"x":1483740000000,"y":72},{"x":1483826400000,"y":88},{"x":1483912800000,"y":68},{"x":1483999200000,"y":32},{"x":1484085600000,"y":48},{"x":1484172000000,"y":115},{"x":1484258400000,"y":47},{"x":1484344800000,"y":88},{"x":1484431200000,"y":20},{"x":1484517600000,"y":108},{"x":1484604000000,"y":41},{"x":1484690400000,"y":52},{"x":1484776800000,"y":101},{"x":1484863200000,"y":60},{"x":1484949600000,"y":96},{"x":1485036000000,"y":65},{"x":1485122400000,"y":42},{"x":1485208800000,"y":118},{"x":1485295200000,"y":106},{"x":1485381600000,"y":50},{"x":1485468000000,"y":92},{"x":1485554400000,"y":101},{"x":1485640800000,"y":93},{"x":1485727200000,"y":94},{"x":1485813600000,"y":43},{"x":1485900000000,"y":116},{"x":1485986400000,"y":24},{"x":1486072800000,"y":105},{"x":1486159200000,"y":80},{"x":1486245600000,"y":29},{"x":1486332000000,"y":72},{"x":1486418400000,"y":52},{"x":1486504800000,"y":116},{"x":1486591200000,"y":40},{"x":1486677600000,"y":67},{"x":1486764000000,"y":27},{"x":1486850400000,"y":86},{"x":1486936800000,"y":67},{"x":1487023200000,"y":0},{"x":1487109600000,"y":0},{"x":1487196000000,"y":0},{"x":1487282400000,"y":0},{"x":1487368800000,"y":0},{"x":1487455200000,"y":0},{"x":1487541600000,"y":0},{"x":1487628000000,"y":0},{"x":1487714400000,"y":0},{"x":1487800800000,"y":0},{"x":1487887200000,"y":0},{"x":1487973600000,"y":0},{"x":1488060000000,"y":0},{"x":1488146400000,"y":0},{"x":1488232800000,"y":0},{"x":1488319200000,"y":0},{"x":1488405600000,"y":0},{"x":1488492000000,"y":11}]}]
```

brylie commented:
Here is a revised diagram:

![screenshot_20170307_133407](https://cloud.githubusercontent.com/assets/17307/23654746/c6f86adc-033a-11e7-9dd9-4e970966325a.png)

From your understanding, what is the point I am trying to make here?

brylie commented:
Lets conduct a couple more **server-side aggregation** experiments, as mentioned above:

- 3,000 - 5,000 records
- 10,000 records
- 50,000 records
- 100,000

That way we can get more insight into how the *aggregation* step might vary as a function of the number of records.

frenchbread commented:
> From your understanding, what is the point I am trying to make here?

@brylie Isn't it "measuring time aggregation takes"? Since aggregation is done on elasticsearch side, search execution metric is exposed in the response object (`3` milliseconds in this case):

![screen shot 2017-03-07 at 14 47 15](https://cloud.githubusercontent.com/assets/2122679/23657076/178933fa-0345-11e7-99e9-1973bafa5a36.png)

frenchbread commented:
### Update

I've modified the server code and moved ElasticSearch client initialisation out of the Meteor method,  this way it saves around `20` milliseconds more.

brylie commented:
> Since aggregation is done on elasticsearch side, search execution metric is exposed in the response object

Thanks for the example. 3ms aggregation with 3,000 records is quite fast. Can we try a couple of experiments with larger data, say 10,000 or 50,000 record aggregation?

frenchbread commented:
@brylie I'm on it. 🚀

frenchbread commented:
Generated sample data & tested request/aggregation time, collected metrics & saved to `docs/request-aggregation-metrics.csv`

### Results

As seen in the CSV file, growing amount of analytics data for **fixed amount of days** (in this case 66 days ~ 2 months) does not slow down request/aggregation time. In fact, referring to numbers speed is even increasing:

| Items | Server -> ES request (ms) | ES search execution (ms) |
| ------ | ----------------------------------- | -------------------- |
| 13k | 42.9 | 3.9 |
| 20k | 30 | 3.7 |
| 50k | 28.7 | 2.3 |

frenchbread commented:
As of now, for this PR provided metrics is enough I suppose. I've created related issue #8 for testing within wider date ranges (larger amount of days).

Nazarah commented:
@frenchbread : great work done. I am really liking the simple look of chart here.
A few improvement suggestions:
* showing the metrics that we are meazuring in X-axis and Y-Axis along with units.
* It would have been great to remove the background grid from the graph so only the plot is seen. But if the graphs are built-in by default that way, we can skip it.

A few queries
* Are we considering any hourly/daily/weekly visualization of data here?
* From EMQTT schema, what metrics we wold be considering to show in visualization as part of API usage monitoring?
* Will we be considering Hourly/Daily/Weekly usage visualizations?

To me the client side monitoring visualization graph is looking more attention grabbing and interesting
https://cloud.githubusercontent.com/assets/2122679/23654223/5d310890-0338-11e7-8aec-5ba59a9ca50e.png
Can we somehow incorporate this sort of visualizations with EMQTT?

frenchbread commented:
@Nazarah Thanks for you comments. I've updated def. of done.

> Are we considering any hourly/daily/weekly visualization of data here?

Yes, it is on the way.

> From EMQTT schema, what metrics we wold be considering to show in visualization as part of API usage monitoring?

That is still not decided. 

> Can we somehow incorporate this sort of visualizations with EMQTT?

That's good idea. What exactly are you referring to?

We could actually divide the current (general overview) chart that we have into multiple smaller ones where each chart would represent usage for specific log type (e.g. `on_client_connected`, `on_client_subscribed` etc.) and color them differently.

How does that sound? @bajiat  @brylie @Nazarah

brylie commented:
> From EMQTT schema, what metrics we wold be considering to show in visualization as part of API usage monitoring?

@frenchbread can you give us some examples of the types of metrics we might expect? E.g. the name of one or more metric(s) and the data type(s). That way we can at least start sketching some ideas.

frenchbread commented:
Here they are https://github.com/apinf/emq-analytics-dashboard/blob/master/dashboard/DATA.md#type-string-possible-options

frenchbread commented:
Added granularity filter:

#### By month

![screen shot 2017-03-09 at 14 33 51](https://cloud.githubusercontent.com/assets/2122679/23750500/f1e63fac-04d5-11e7-95ff-255203a282dd.png)

#### By week

![screen shot 2017-03-09 at 14 33 57](https://cloud.githubusercontent.com/assets/2122679/23750505/f89f36f0-04d5-11e7-87c1-62ba6c0039e6.png)

#### By day

![screen shot 2017-03-09 at 14 34 06](https://cloud.githubusercontent.com/assets/2122679/23750511/fee96ca6-04d5-11e7-88e4-cb29076782e5.png)

#### By hour

![screen shot 2017-03-09 at 14 34 13](https://cloud.githubusercontent.com/assets/2122679/23750518/050b3dee-04d6-11e7-9485-9f462b3ed0fc.png)

frenchbread commented:
This is what I meant by 

> We could actually devide the current (general overview) chart that we have into multiple smaller ones where each chart would represent usage for specific log type (e.g. on_client_connected, on_client_subscribed etc.) and color them differently.

![screen shot 2017-03-09 at 16 40 55](https://cloud.githubusercontent.com/assets/2122679/23754870/33bc1a6c-04e7-11e7-931d-edbe59499403.png)

Are we going to keep this (& I commit changes here)?

frenchbread commented:
There are some drawbacks here: 
- When selecting a wide dates range (e.g. 2 years) & granularity `hour` - dashboard freezes

In other cases works fine

brylie commented:
Why does the dashboard freeze with large date ranges?

frenchbread commented:
I assume it happens due rendering data for large amount of days (> 600 days).

frenchbread commented:
`Idea`: we can create use cases for filter queries when dashboard freezes, hide chart and show placeholder message e.g. "Too much data to show" or "Select smaller dates range"..

brylie commented:
> I assume it happens due rendering data for large amount of days (> 600 days).

How can you test that hypothesis?

brylie commented:
Another idea is to have 'default' granularity settings. E.g. when selecting two years span, defaulting the query to monthly (or daily).

frenchbread commented:
Removed grid lines & changed datepicker to `pikaday` library.

### TODO:
- Move `log type` filter to drop down menu

brylie commented:
Is this ready for review, or will it be under construction a bit longer?

frenchbread commented:
@brylie This is not yet ready.

frenchbread commented:
Moved filtering by message type to its own select-picker.

![screen shot 2017-03-14 at 16 26 33](https://cloud.githubusercontent.com/assets/2122679/23905403/a041d97a-08d3-11e7-849c-9c2c25d0cc5d.png)

Please review.

brylie commented:
Woohoo! 😀

frenchbread commented:
@brylie Could you merge this? If you are going to test it, I can send you ES host with sample data.

brylie commented on core/lib/client/pikaday.css:
```diff
@@ -0,0 +1,222 @@
+@charset "UTF-8";
```
Is it difficult/impossible to import this CSS from the NPM package?

brylie commented on dashboard/client/index.js:
```diff
@@ -1,137 +1,146 @@
-    },
-  };
+  instance.updateQuery = () => {
+    const from = FlowRouter.getQueryParam('from') || moment().subtract(1, 'month').format('YYYY-MM-DD')
```
Why do we duplicate this here? The FlowRouter `from`, `to`, `granularity`, and `type` query parameters should be set onRender, correct?

brylie replied:
Is there a case where this code is failing without the `||`?

frenchbread replied:
There is a case when URL does not contain `from` or `to` params, but we still need to have default date range to request data for.

brylie commented on dashboard/client/index.js:
```diff
@@ -1,137 +1,146 @@
+    const type = FlowRouter.getQueryParam('type') || ''
+
+    // For metics collection
+    // const a = moment(from, 'YYYY-MM-DD')
```
Clean up commented code.

brylie commented on dashboard/client/index.js:
```diff
@@ -1,137 +1,146 @@
+    // const b = moment(to, 'YYYY-MM-DD')
+    // console.log(`${b.diff(a, 'days')} days shown.`)
+
+    const q = instance.opts.body.query.bool.must
```
What does `q` refer to? Lets give this variable a semantic name, so we can [avoid mental mapping](https://github.com/ryanmcdermott/clean-code-javascript#avoid-mental-mapping).

brylie commented on dashboard/server/methods.js:
```diff
@@ -1,18 +1,84 @@
 Meteor.methods({
-  getEsData (opts) {
+  getAggr (opts) {
+    // const searches = _.map(types, type => {
```
Clean up commented code.

brylie commented on dashboard/server/methods.js:
```diff
@@ -1,18 +1,84 @@
+    const startTime = new Date().getTime()
+    return client.search(opts).then(res => {
+      const took1 = new Date().getTime() - startTime
+      console.log(`--------------`)
```
Clean up console logs.

brylie commented on dashboard/server/methods.js:
```diff
@@ -1,18 +1,84 @@
-    }, (err) => {
-      throw new Meteor.Error(err.message);
-    });
+      return [
```
This function name `getAggr` basically says 'get aggregate data', but we are adding chart style parameters here. [Functions should do one thing](https://github.com/ryanmcdermott/clean-code-javascript#functions-should-do-one-thing), so this function should only return the aggregate data without any chart style rules.

brylie commented on dashboard/server/methods.js:
```diff
@@ -1,18 +1,84 @@
 
 Meteor.methods({
-  getEsData (opts) {
+  getAggr (opts) {
```
While this function name pretty much describes what it does, it should also be [meaningful and pronouncable](https://github.com/ryanmcdermott/clean-code-javascript#use-meaningful-and-pronounceable-variable-names). `getAggr` is *almost* meaningful and *sort-of* pronouncable, but would be much better as `getAggregateData`.

brylie commented on dashboard/server/startup.js:
```diff
@@ -3,35 +3,86 @@
 // import moment from 'moment'
```
Clean up commented code.

brylie commented on dashboard/server/startup.js:
```diff
@@ -3,35 +3,86 @@
 //
 // const diff = dateNow.diff(dateStart, 'days')
 //
+// const types = [
```
Clean up commented code.

brylie commented on dashboard/server/startup.js:
```diff
@@ -3,35 +3,86 @@
 //       }
 //     }
+//
+//     if (randomType === 'on_client_subscribe') {
```
Clean up commented code.

brylie commented on docs/request-aggregation-metrics.csv:
```diff
@@ -0,0 +1,33 @@
+Items count,Days amount,Server → ES requrest (ms),ES search exec (ms)
```
Consider moving this to the `apinf/docs` repository, to a folder containing research and testing related to your EMQ work.

brylie commented:
This looks really good from a UI perspective! I made some suggestions on how to improve the code style, so our code conforms to our [developer expectations](https://github.com/apinf/platform/blob/develop/CONTRIBUTING.md#code-quality).

phanimahesh commented:
Was this checked with live ES data? If not, @frenchbread let's do a quick check with live data once. First thing tomorrow morning, unless you are occupied. We exchanged docs on event format but still.

frenchbread commented:
Added suggested changes. Please review

frenchbread commented:
After new elasticsearch plugin for emq is deployed, it would require to refactor a little bit in order to fit updated analytics logs schema. But this could be done in separate PR

frenchbread commented:
@phanimahesh As we discussed in the chat, yes, dashboard has been tested with live/real ES data. For purposes of testing (with wider date-ranges & collecting metrics of loading/rendering) some dummy data was added in addition to 'real' generated by the plugin.

brylie commented on dashboard/client/index.js:
```diff
@@ -41,8 +41,18 @@ Template.dashboard.onCreated(function () {
+  instance.getChartData = opts => new Promise((resolve, reject) => {
+    Meteor.call('getAggregatedData', opts, (err, res) => {
+      if (err) reject(err)
+      Meteor.call('parseDataForNvd', res, (err1, res1) => {
```
Should this be a Meteor method, or just a client side funciton? I am asking, be cause Meteor methods are mainly server-side (with client side stub). That means, in this scenario, we are getting aggregated data from the server and then sending it back to the server to be parsed and returned again.

frenchbread replied:
This is a server-side method created to decrease the load on client-side (browser).

> That means, in this scenario, we are getting aggregated data from the server and then sending it back to the server to be parsed and returned again.

Right.

brylie replied:
I think we should **reduce the number of round trips**. If we want to decrease the client load by performing a second aggregation on the server, then **send the data only after it has been aggregated properly**.

brylie commented on dashboard/client/index.js:
```diff
@@ -41,8 +41,18 @@ Template.dashboard.onCreated(function () {
+  })
+
+  instance.parseDataForNvd = data => new Promise((resolve, reject) => {
+    Meteor.call('parseDataForNvd', data, (err, res) => {
```
Again, it seems like this parsing step (preparing it for NVD3 should either:

- happen as a sub-function call in the `getChartData` server method
- happen as a client side parsing function

It seems inefficient that we would get data from the server and send it back to the server for further parsing.

brylie commented on dashboard/client/index.js:
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
+    const range = _.find(mustQuery, obj => typeof obj.range !== 'undefined')
     if (range) {
-      _.remove(q, _.find(q, obj => typeof obj.range !== 'undefined'))
+      _.remove(mustQuery, _.find(mustQuery, obj => typeof obj.range !== 'undefined'))
```
This line of code deserves a human friendly comment.

brylie commented on dashboard/client/index.js:
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
+    const mustQuery = instance.opts.body.query.bool.must
 
-    const range = _.find(q, obj => typeof obj.range !== 'undefined')
+    const range = _.find(mustQuery, obj => typeof obj.range !== 'undefined')
```
Add comment here describing this line of code in plain language.

brylie commented on dashboard/client/index.js:
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
     }
 
-    const match = _.find(q, obj => typeof obj.match !== 'undefined')
+    const match = _.find(mustQuery, obj => typeof obj.match !== 'undefined')
```
Add comment.

brylie commented on dashboard/client/index.js:
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
+    const match = _.find(mustQuery, obj => typeof obj.match !== 'undefined')
     if (match) {
-      _.remove(q, _.find(q, obj => typeof obj.match !== 'undefined'))
+      _.remove(mustQuery, _.find(mustQuery, obj => typeof obj.match !== 'undefined'))
```
Add comment.

frenchbread commented:
Added comments

frenchbread commented:
Removed second meteor method. Please :mag_right:
